[ASTContext] Add setter for PreModuleImportCallback

chelcassanova opened the pull request:
`PreModuleImportCallback` is a variable for a callback function called before module importing in the Swift compiler. It is currently being used by LLDB to report progress updates on importing Swift modules in LLDB. This commit adds a setter for `PreModuleImportCallback` for greater flexibility in LLDB and removes the reference to this variable in the constructor.

Related to https://github.com/apple/llvm-project/pull/7769

rdar://105286354

chelcassanova committed:
[ASTContext] Add setter for LLDB callback

`PreModuleImportCallback` is a variable for a callback function used
specifically by LLDB to report progress updates on importing Swift
modules in LLDB. This commit adds a setter for `PreModuleImportCallback`
for greater flexibility in LLDB and removes the reference to this
variable in the constructor.

Related to https://github.com/apple/llvm-project/pull/7769

rdar://105286354

medismailben commented on include/swift/AST/ASTContext.h:
```diff
@@ -239,8 +239,8 @@ class ASTContext final {
-      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutBackend = nullptr,
-      std::function<bool(llvm::StringRef, bool)> PreModuleImportCallback = {});
+      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutBackend = nullptr
+      );
```
Interesting formatting. Is that really `clang-format`'s output ?

chelcassanova replied:
Yeah it is, other than removing the callback in the constructor I didn't change anything else here.

medismailben commented on lib/AST/ASTContext.cpp:
```diff
@@ -715,6 +714,10 @@ ASTContext::~ASTContext() {
 }
 
+void ASTContext::SetPreModuleImportCallback(std::function<bool(llvm::StringRef ModuleName, bool IsOverlay)> callback) {
+  ASTContext::PreModuleImportCallback = callback;
```
Why do you need to qualify `PreModuleImportCallback` here ?

chelcassanova replied:
I don't, but I think I originally thought that `PreModuleImportCallback` was static even though it's not so I can remove the qualification.

medismailben commented:
FWIW, although this callback is used mainly as an LLDB module loading progress callback, it could also be used for/by other things. I'd retitle this PR to be more generic.

chelcassanova commented:
@swift-ci test

chelcassanova commented:
Good point about `PreModuleImportCallback` being generic. Since LLDB was the only one use it I thought it was created specifically for LLDB, I'll update the PR description to reflect this.

adrian-prantl commented on include/swift/AST/ASTContext.h:
```diff
@@ -294,6 +294,8 @@ class ASTContext final {
   /// OutputBackend for writing outputs.
   llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutputBackend;
 
+  void SetPreModuleImportCallback(std::function<bool(llvm::StringRef ModuleName, bool IsOverlay)> callback);
```
Doxygen comment?

chelcassanova replied:
Yes, I'll add one here

chelcassanova committed:
Change callback return type and add Doxygen comment

Sets the `PreModuleImportCallback`'s setter return type to void to
account for the lambda functions in LLDB's progress reporting not
actually using its return value. Also adds a Doxygen comment for `SetPreModuleImportCallback`

chelcassanova commented:
@swift-ci test

chelcassanova commented:
@swift-ci smoke test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci  test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please smoke test

augusto2112 commented on lib/AST/ASTContext.cpp:
```diff
@@ -715,6 +714,10 @@ ASTContext::~ASTContext() {
   getImpl().~Implementation();
 }
 
+void ASTContext::SetPreModuleImportCallback(std::function<void(llvm::StringRef ModuleName, bool IsOverlay)> callback) {
```
Was this line `clang-format`ted? Looks pretty long

chelcassanova replied:
It should've been, but I can also shorten it myself if `clang-format` is ignoring it for some reason

augusto2112 replied:
If `clang-format` says it's ok then it's fine

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

chelcassanova committed:
fixup! Change callback return type and add Doxygen comment

Run `clang-format`

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test

chelcassanova commented:
https://github.com/apple/llvm-project/pull/7769
@swift-ci please test
